rpc/src/v1/helpers/light_fetch.rs

```diff
@@ -438,6 +435,154 @@ impl LightFetch {
 			None => Box::new(future::err(errors::network_disabled())) as Box<Future<Item = _, Error = _> + Send>
 		}
 	}
+
+	fn headers_range_by_block_id(
+		&self,
+		from_block: BlockId,
+		to_block: BlockId,
+		max: u64
+	) -> impl Future<Item = Vec<encoded::Header>, Error = Error> {
+		let fetch_hashes = [from_block, to_block].iter()
+			.filter_map(|block_id| match block_id {
+				BlockId::Hash(hash) => Some(hash.clone()),
+				_ => None,
+			})
+			.collect::<Vec<_>>();
+
+		let best_number = self.client.chain_info().best_block_number;
+
+		let fetcher: Self = self.clone();
+		self.headers_by_hash(fetch_hashes).and_then(move |mut header_map| {
+			let (from_block_num, to_block_num) = {
+				let block_number = |id| match id {
+					&BlockId::Earliest => 0,
+					&BlockId::Latest => best_number,
+					&BlockId::Hash(ref h) =>
+						header_map.get(h).map(|hdr| hdr.number())
+						.expect("from_block and to_block headers are fetched by hash; this closure is only called on from_block and to_block; qed"),
+					&BlockId::Number(x) => x,
+				};
+				(block_number(&from_block), block_number(&to_block))
+			};
+
+			if to_block_num < from_block_num {
+				// early exit for "to" block before "from" block.
+				return Either::A(future::ok(Vec::new()));
+			} else if to_block_num - from_block_num >= max {
+				return Either::A(future::err(errors::request_rejected_param_limit()));
+			}
+
+			let to_header_hint = match to_block {
+				BlockId::Hash(ref h) => header_map.remove(h),
+				_ => None,
+			};
+			let headers_fut = fetcher.headers_range(from_block_num, to_block_num, to_header_hint);
+			Either::B(headers_fut.map(move |headers| {
+				// Validate from_block if it's a hash
+				let last_hash = headers.last().map(|hdr| hdr.hash());
+				match (last_hash, from_block) {
+					(Some(h1), BlockId::Hash(h2)) if h1 != h2 => Vec::new(),
+					_ => headers,
+				}
+			}))
+		})
+	}
+
+	fn headers_by_hash(&self, hashes: Vec<H256>) -> impl Future<Item = HashMap<H256, encoded::Header>, Error = Error> {
+		let mut refs = HashMap::with_capacity(hashes.len());
+		let mut reqs = Vec::with_capacity(hashes.len());
+
+		for hash in hashes {
+			refs.entry(hash).or_insert_with(|| {
+				if let Some(header) = self.client.block_header(BlockId::Hash(hash)) {
+					header.into()
+				} else {
+					reqs.push(request::HeaderByHash(hash.into()).into());
+					HeaderRef::Unresolved(reqs.len() - 1, hash.into())
+				}
```

Comment:
It seems kind of roundabout to wrap a value in an enum then unwrap the result because we know how that branch is handled in a helper function. I figured it was preferable here to avoid the `expect` and accept some code duplication here, but I'll change it if you really prefer.

As for the deduping, this method does not require the input hash vector to be sorted (or of length <= 2), so the `or_insert_with` is the easiest way to dedup IMO, rather than doing a sort first.

Reply:
Even though I agree that [sometimes](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction) duplication is better, IMHO in this case using `make_header_requests` makes code more readable and DRY (e.g. we can change the logic in one place). As for possible performance impact, I think if `make_header_requests` is inlined, then the impact should be [negligible](https://godbolt.org/g/fLTaeS).

As for deduping, since this is a private method, we can control the requirements. But I'm fine with leaving it as is.